File: 1760-minimum-limit-of-balls-in-a-bag/1760-minimum-limit-of-balls-in-a-bag.java
import java.util.*;

class Solution {
    public int minimumSize(int[] nums, int maxOperations) {
        int left = 1, right = Arrays.stream(nums).max().getAsInt(), result = -1;

        while (left <= right) {
            int mid = (left + right) / 2;
            if (canDivide(nums, maxOperations, mid)) {
                result = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return result;
    }

    private boolean canDivide(int[] nums, int maxOperations, int mid) {
        int operations = 0;
        for (int bag : nums) {
            operations += (bag - 1) / mid;
        }
        return operations <= maxOperations;
    }
}

File: 1760-minimum-limit-of-balls-in-a-bag/1760-minimum-limit-of-balls-in-a-bag.py
class Solution:
    def minimumSize(self, nums: list[int], maxOperations: int) -> int:
        def canDivide(mid):
            operations = 0
            for bag in nums:
                operations += (bag - 1) // mid
            return operations <= maxOperations

        left, right = 1, max(nums)
        result = -1

        while left <= right:
            mid = (left + right) // 2
            if canDivide(mid):
                result = mid
                right = mid - 1
            else:
                left = mid + 1

        return result
